config/config.go

```diff
@@ -1015,6 +1047,70 @@ func configFileHasDependencyBlock(configPath string, terragruntOptions *options.
 	return false, nil
 }
 
+const (
+	MetadataTerraform                   = "terraform"
+	MetadataTerraformBinary             = "terraform_binary"
+	MetadataTerraformVersionConstraint  = "terraform_version_constraint"
+	MetadataTerragruntVersionConstraint = "terragrunt_version_constraint"
+	MetadataRemoteState                 = "remote_state"
+	MetadataDependencies                = "dependencies"
+	MetadataDependency                  = "dependency"
+	MetadataDownloadDir                 = "download_dir"
+	MetadataPreventDestroy              = "prevent_destroy"
+	MetadataSkip                        = "skip"
+	MetadataIamRole                     = "iam_role"
+	MetadataIamAssumeRoleDuration       = "iam_assume_role_duration"
+	MetadataIamAssumeRoleSessionName    = "iam_assume_role_session_name"
+	MetadataInputs                      = "inputs"
+	MetadataLocals                      = "locals"
+	MetadataGenerateConfigs             = "generate"
+	MetadataRetryableErrors             = "retryable_errors"
+	MetadataRetryMaxAttempts            = "retry_max_attempts"
+	MetadataRetrySleepIntervalSec       = "retry_sleep_interval_sec"
+)
+
+// AddFieldMetadata persist "generic" metadata
+func (conf *TerragruntConfig) MetadataWithType(fieldType, fieldName string, m map[string]interface{}) {
+	if conf.FieldsMetadata == nil {
+		conf.FieldsMetadata = map[string]map[string]interface{}{}
+	}
+
+	field := fmt.Sprintf("%s-%s", fieldType, fieldName)
+
+	metadata, found := conf.FieldsMetadata[field]
+	if !found {
+		metadata = make(map[string]interface{})
+	}
+	for key, value := range m {
+		metadata[key] = value
+	}
+	conf.FieldsMetadata[field] = metadata
+}
+
+func (conf *TerragruntConfig) Metadata(fieldName string, m map[string]interface{}) {
```

Comment:
NIT: I think it would be slightly clearer to name these `MetadataXXX` methods something like `SetMetadataXXX`. That way, you know it's a method to mutate the underlying object. Or perhaps even `SetFieldMetadata` since you have a `GetFieldMetadata` method.